### DISCUSSION TITLE
react hook form testing isValid always false

### DISCUSSION THREAD

**Discussion by Santosl2** [react-hook-form/react-hook-form#9699] [Q&A] [answered]
Hi everyone, recently i tried testing my Form, and my submit button has attribute `disabled` and the same receive the react hook form `isValid` prop.

But the case is: the prop `isValid` not change

I using:
- Next JS
- Zod Schema
- React Select

## Details:

i mocked my react-select component and the result is equal this

![image](https://user-images.githubusercontent.com/67132916/210298829-ce3d2baf-37e0-4a0b-b1a6-636db529fb20.png)

# Snippets

![image](https://user-images.githubusercontent.com/67132916/210298944-1d6f0e4b-29f2-4738-977b-68e133cc4ca0.png)

### the test
![image](https://user-images.githubusercontent.com/67132916/210298683-f2201167-ad1f-4689-8041-0fa35874a6aa.png)

**Comment by leapful** [comment]
@Santosl2 

Please share a codesandbox url to re-produce the issue.

**Reply by Santosl2** [reply]
Yes dude.

**Reply by Santosl2** [reply]
Is here https://codesandbox.io/s/funny-yonath-m1nxu2?file=/pages/index.tsx.  Just run `npm run test` to re-produce the issue

**Comment by Santosl2** [comment]
So... someone can help me?

**Comment by leapful** [comment, marked as answer]
@Santosl2 

Your jest.setup.ts file for finding option value is not correct because option value is number but data from select element is string. It should be:

jest.setup.ts, line 15
```
const option = options.find(
        (op) => `${op.value}` === event.currentTarget.value
      );
```

Furthermore, `formState.isValid` is a proxy object therefore it will be updated in next re-rendering, so a custom wait should be applied in order to get update formState value.

index.test.ts
```
describe("Home", () => {
  it("button must be enabled when user select your option", async () => {
    render(<Home />);

    expect(screen.getByText("Continue")).toBeDisabled();

    const input = screen.getByTestId("select");

    fireEvent.change(input, {
      target: {
        value: 2
      }
    });

    // Add custom wait to get correct form state in next re-rendering
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(screen.getByText("Continue")).toBeEnabled();
    });
  });
});

```

**Reply by Santosl2** [reply]
WOWW! How did i not notice this? LOL

Thx dude!!

I use your solution, but with some refactors.

in jest.setup.ts, line 15
```js
  const option = options.find(
        (op) => op.value == event.currentTarget.value
      );
```

I only changed the strict equality operator for the equality operator.

And in my index.test.ts, i used RTL`s act method.. It was something like:

```js
describe("Home", () => {
  it("button must be enabled when user select your option", async () => {
    render(<Home />);

    expect(screen.getByText("Continue")).toBeDisabled();

    const input = screen.getByTestId("select");

    await act(async () => {
      fireEvent.change(input, {
        target: {
          value: 2
        }
      });
    });

    expect(screen.getByText("Continue")).toBeEnabled();
  });
});
```

So men, thank you very much!
